Allow cluster to start when standby host is down

jmcatamney opened the pull request:
Previously, gpstart could not start the cluster if a standby master host was
configured but currently down.  In order to check whether the standby was
supposed to be the acting master (and prevent the master from being started if
that was the case), gpstart needed to access the standby host to retrieve the
TimeLineID of the standby, and if the standby host was down the master would not
start.

This commit modifies gpstart to assume that the master host is the acting master
if the standby is unreachable, so that it never gets into a state where neither
the master nor the standby can be started.

---

A bit of previous PR discussion [here](https://github.com/greenplum-db/gpdb/pull/9513), as this was first addressed on 5X and a PR opened there; any changes made to this PR will be applied to that one as well.

berlin-ab commented on gpMgmt/bin/gppylib/test/unit/test_unit_gpstart.py:
```diff
@@ -205,6 +206,24 @@ def test_prepare_segment_start_returns_up_and_down_segments(self):
 
+    @patch("gppylib.commands.pg.PgControlData.run")
+    @patch("gppylib.commands.pg.PgControlData.get_value", return_value="2")
+    def test_fetch_tli_when_standby_is_accessible(self, mock1, mock2):
```
nit: the test descriptions could be improved to include a "what to expect and why" at a high level

berlin-ab commented on gpMgmt/bin/gpstart:
```diff
@@ -231,8 +231,11 @@ class GpStart:
+        try:
+            controldata.run(validateAfter=True)
+            return int(controldata.get_value("Latest checkpoint's TimeLineID"))
+        except: # Can't access the standby, so return a value that won't lead to standby promotion
```
This made me think of: 

https://github.com/benchristel/refactoring-workshop/blob/master/docs/refactorings/move-telic-comment-to-test.md

berlin-ab replied:
And the related smell: 

https://github.com/benchristel/refactoring-workshop/blob/master/docs/smells/telic-comment-in-code.md

marksliva replied:
We moved this to the user confirmation dialog.

jchampio commented:
> A bit of previous PR discussion here

Should I migrate my comments, or are you already working on those?

jchampio commented:
> This commit modifies gpstart to assume that the master host is the acting master
if the standby is unreachable, so that it never gets into a state where neither
the master nor the standby can be started.

Doesn't this reintroduce the possibility of a split-brain system, which the prior behavior was trying to prevent?

jmcatamney commented:
> Should I migrate my comments, or are you already working on those?

I fixed the `ExecutionError` bit and added more detail to the commit message, so I think that covers the previous comments.

> Doesn't this reintroduce the possibility of a split-brain system, which the prior behavior was trying to prevent?

Good point.  We'll chat with PM about how we want to handle that situation.

jmcatamney commented:
After talking with PM and Server, we believe the safest way to handle a split-brain situation is to trust the user to know whether it's safe to start the master while the standby host is down, so we've added a warning message and prompt to that effect.

dkrieger-pvtl commented on gpMgmt/bin/gpstart:
```diff
@@ -225,14 +218,32 @@ class GpStart:
-        return int(controldata.get_value("Latest checkpoint's TimeLineID"))
+        try:
+            controldata.run(validateAfter=True)
+            return int(controldata.get_value("Latest checkpoint's TimeLineID"))
```
nit: the `PgControlData` command is called `Latest checkpoint's TimeLineID` AND the key of the value we want from that file is also called the same thing `Latest checkpoint's TimeLineID`.   But the command name doesn't have to be a key value, and perhaps should be something else like `obtain pg_controldata timelineID` or such.  

I can see not wanting to change this in 6X, so you could leave it as is, but I still find the repeat of the same string a bit awkward when scanning the code.

kalensk replied:
Added some clarity to the PgControlData command descriptions with `fetching pg_controldata locally` and `fetching pg_controldata remotely`.

asimrp commented:
That means PR #9513 will be closed and this change will be back patched to 5X, correct?  If so, it will be good to close the previous PR.

The added warning makes sense, I feel quite relieved now.  However, the question remains - what is the purpose of `--skip_standby_check` option?

jmcatamney commented:
> The added warning makes sense, I feel quite relieved now. However, the question remains - what is the purpose of `--skip_standby_check option`?

That option is hidden (which is why the user who ran into this bug didn't try it), but I assume it's there for Support to use if they expect the standby host to be down and are confident that the check can be skipped, whereas the logic added in this PR can catch a case where the standby host is down unexpectedly.

We talked with Bharath and have decided to keep the check around for now for those who find it useful, since it can skip the added confirmation prompt and so is easier to script on the Support side.

jchampio commented on gpMgmt/bin/gpstart:
```diff
@@ -225,14 +218,32 @@ class GpStart:
+        except base.ExecutionError:
+            logger.warning("Standby host is unreachable, cannot determine whether the standby is currently acting as the master.")
+            logger.warning("Proceed only if you are certain that the standby is not acting as the master.")
+            if not userinput.ask_yesno(None, "\nContinue with startup", 'N'):
```
What happens if the user has scripted a `gpstart -a`? Does `ask_yesno()` assume that a "no" is answered if we're not currently connected to a tty?

marksliva replied:
Good catch. Pushed a change.

jchampio commented on gpMgmt/bin/gpstart:
```diff
@@ -225,14 +218,32 @@ class GpStart:
+            logger.warning("Standby host is unreachable, cannot determine whether the standby is currently acting as the master.")
+            logger.warning("Proceed only if you are certain that the standby is not acting as the master.")
+            if not userinput.ask_yesno(None, "\nContinue with startup", 'N'):
+                self.shutdown_master_only()
```
I realize that the current code isn't written to be exception-safe, but the multiple points of cleanup that we have make me think that we badly need it. Is this something we could have moved into a finally block up above, instead?

jchampio replied:
Looks like there are additional risks -- the current `try` code isn't scoped to handle exception-safety correctly anyway. It's painful but I withdraw my request.

asimrp commented:
One more loud thought - how about recording (e.g. create a file in master data directory) a user's decision to move forward even when standby state could not be determined?  It will serve as a clue to finding root cause, should a split brain situation indeed arise.  The decision is currently logged in gpAdminLogs but the logs can be easily lost.

ashwinstar commented:
> On Feb 5, 2020, at 7:46 PM, Asim R P <notifications@github.com> wrote:
> 
> ﻿
> One more loud thought - how about recording (e.g. create a file in master data directory) a user's decision to move forward even when standby state could not be determined? It will serve as a clue to finding root cause, should a split brain situation indeed arise. The decision is currently logged in gpAdminLogs but the logs can be easily lost.
> 


Nice thought, though seems too too heavy lifting for purpose, value and number of times we will use the info. gpAdminLogs seems fine, if any problem it will show up pretty soon not after days, so will easily have access to gpAdminLogs.

asimrp commented on gpMgmt/bin/gpstart:
```diff
@@ -225,14 +218,32 @@ class GpStart:
+            controldata.run(validateAfter=True)
+            return int(controldata.get_value("Latest checkpoint's TimeLineID"))
+        except base.ExecutionError:
+            logger.warning("Standby host is unreachable, cannot determine whether the standby is currently acting as the master.")
```
Will the `ExecutionError` be reported in the logs by the time we get here?  If not, it will be good to include string representation of it in this message.

marksliva replied:
Fixed.
